src/constraint_system/boolean.rs

```diff
@@ -50,6 +51,51 @@ where
 
         a
     }
+
+    /// A gate which outputs a variable whose value is 1 if
+    /// the input is 0 and whose value is 0 otherwise
+    pub fn is_zero_gate(&mut self, a: Variable) -> Variable {
+        //Get relevant field values
+        let a_value = self.variables.get(&a).unwrap();
+        let y_value = a_value.inverse().unwrap_or_else(E::Fr::one);
+        //This has value 1 if input value is zero, value 0 otherwise
+        let b_value = E::Fr::one() - *a_value * y_value;
+
+        let y = self.add_input(y_value);
+        let b = self.add_input(b_value);
+        let zero = self.zero_var();
+
+        // Enforce constraints. The constraint system being used here is
+        // a * y + b - 1 = 0
+        // a * b = 0
+        // where y is auxiliary and b is the boolean (a == 0).
+        let _a_times_b = self.arithmetic_gate(|gate| {
+            gate.witness(a, b, Some(zero)).mul(E::Fr::one())
+        });
+
+        let _constraint1 = self.arithmetic_gate(|gate| {
+            gate.witness(a, y, Some(zero))
+                .mul(E::Fr::one())
+                .fan_in_3(E::Fr::one(), b)
+                .constant(-E::Fr::one())
+        });
+
+        b
+    }
+
+    /// A gate which outputs a variable whose value is 1 if the
+    /// two input variables have equal values and whose value is 0 otherwise.
+    pub fn is_eq_gate(&mut self, a: Variable, b: Variable) -> Variable {
```

Comment:
There's already a fn that does this in this repo.

Check: https://github.com/ZK-Garage/plonk/blob/9e7c26539816bf33873cbeb51672da98f6a50e85/plonk-core/src/constraint_system/composer.rs#L327

So this can be removed, or you can add this fn in the `composer.rs` module calling it something like: `is_eq_with_output` or something along these lines.

Reply:
Thanks for your comments.  Right, the `assert_equal` fn already exists but does not give the user any output variable.  That is the reason for this PR, so it cannot be removed.  Are you suggesting that I locate it in the composer.rs module instead?

Reply:
Exactly. And that you rename it on a way on which it implies the return of a `Veriable` in opposite to the other fn, :)

Reply:
Okay, all done